tests/gamma/csm_observability_test.py

```diff
@@ -11,24 +11,126 @@
+    f"{PROMETHEUS_HOST}/"
+    "grpc_client_attempt_sent_total_compressed_message_size_bytes/histogram"
+)
+METRIC_CLIENT_ATTEMPT_RCVD = (
+    f"{PROMETHEUS_HOST}/"
+    "grpc_client_attempt_rcvd_total_compressed_message_size_bytes/histogram"
+)
+METRIC_CLIENT_ATTEMPT_DURATION = (
+    f"{PROMETHEUS_HOST}/grpc_client_attempt_duration_seconds/histogram"
+)
+METRIC_CLIENT_ATTEMPT_STARTED = (
+    f"{PROMETHEUS_HOST}/grpc_client_attempt_started_total/counter"
+)
+METRIC_SERVER_CALL_RCVD = (
+    f"{PROMETHEUS_HOST}/"
+    "grpc_server_call_rcvd_total_compressed_message_size_bytes/histogram"
+)
+METRIC_SERVER_CALL_SENT = (
+    f"{PROMETHEUS_HOST}/"
+    "grpc_server_call_sent_total_compressed_message_size_bytes/histogram"
+)
+METRIC_SERVER_CALL_DURATION = (
+    f"{PROMETHEUS_HOST}/grpc_server_call_duration_seconds/histogram"
+)
+METRIC_SERVER_CALL_STARTED = (
+    f"{PROMETHEUS_HOST}/grpc_server_call_started_total/counter"
+)
+HISTOGRAM_CLIENT_METRICS = (
+    METRIC_CLIENT_ATTEMPT_SENT,
+    METRIC_CLIENT_ATTEMPT_RCVD,
+    METRIC_CLIENT_ATTEMPT_DURATION,
+)
+HISTOGRAM_SERVER_METRICS = (
+    METRIC_SERVER_CALL_DURATION,
+    METRIC_SERVER_CALL_RCVD,
+    METRIC_SERVER_CALL_SENT,
+)
+COUNTER_CLIENT_METRICS = (METRIC_CLIENT_ATTEMPT_STARTED,)
+COUNTER_SERVER_METRICS = (METRIC_SERVER_CALL_STARTED,)
+HISTOGRAM_METRICS = HISTOGRAM_CLIENT_METRICS + HISTOGRAM_SERVER_METRICS
+COUNTER_METRICS = COUNTER_CLIENT_METRICS + COUNTER_SERVER_METRICS
+CLIENT_METRICS = HISTOGRAM_CLIENT_METRICS + COUNTER_CLIENT_METRICS
+SERVER_METRICS = HISTOGRAM_SERVER_METRICS + COUNTER_SERVER_METRICS
+ALL_METRICS = HISTOGRAM_METRICS + COUNTER_METRICS
+
+GammaServerRunner = gamma_server_runner.GammaServerRunner
+KubernetesClientRunner = k8s_xds_client_runner.KubernetesClientRunner
+
+
+# This class should represent one TimeSeries object from
+# monitoring_v3.ListTimeSeriesResponse.
+@dataclasses.dataclass(eq=False)
+class MetricTimeSeries:
+    # the metric name
+    name: str
+    # each time series has a set of metric labels
+    metric_labels: dict[str, str]
+    # each time series has a monitored resource
+    resource_type: str
+    # each time series has a set of monitored resource labels
+    resource_labels: dict[str, str]
+    # each time series has a set of data points
+    points: list[monitoring_v3.types.Point]
+
+    @classmethod
+    def from_response(
+        cls,
+        name: str,
+        response: monitoring_v3.types.TimeSeries,
+    ) -> "MetricTimeSeries":
+        return cls(
+            name=name,
+            metric_labels=dict(response.metric.labels),
+            resource_type=response.resource.type,
+            resource_labels=dict(response.resource.labels),
+            points=list(response.points),
+        )
+
+    def pretty_print(self) -> str:
+        metric = dataclasses.asdict(self)
+        # too much noise to print all data points from a time series
+        metric.pop("points")
+        return yaml.dump(metric, sort_keys=True)
```

Comment:
If we go with the presort I suggested above, we may benefit from printing while preserving the top-level fields order (name, resource_type, resource_labels, etc). If so, would be nice to place `resource_type: str` right next to the name

Reply:
Done